posts/2020-10-02-the-tale-about-hibernate-and-temporal-tables.adoc

```diff
@@ -0,0 +1,149 @@
+= The tale about Hibernate and Temporal Tables
+Nataliia Dziubenko <nataliia.dziubenko@lunatech.nl>
+v1.0, 2020-10-02
+:title: The tale about Hibernate and Temporal Tables
+:tags: [hibernate, java, temporal, mssql]
+
+_thanks Willem-Jan Glerum for introducing me to this interesting behaviour_
+
+== Foreword
+
+The project where this problem was encountered, was using Hibernate + Envers and Microsoft Sql Server as the database. The choice of database was mainly driven by requirement to have history data which could be provided by temporal tables. In Java code there was an entity class for history data `VersionedRecord` which was a `@MappedSuperclass` and couple of entities were extending it. And at this moment, when fetching data into these entities from the database, something unexpected happened: Hibernate reused repeatedly only data for the first row, ignoring the rest of the result set! That was caused by Hibernate first-level caching, and we had to look for some trick to overcome this without changing the `VersionedRecord` entity class. The main goal was to make a change as painless as possible and avoid changes in the entity class. Creation of extra entity class just for this purpose seemed like an overkill.
+
+In this topic I will explain what happened on the clean and simple example and also share a trick that we used to solve this, hopefully it will be useful for someone else.
+
+== The experiment
+
+To make the experiment clean, I used only Hibernate and nothing on top of it. So, first of all I created a table with history table like this:
+[source,sql]
+----
+CREATE TABLE Department
+(
+    DeptID INT NOT NULL PRIMARY KEY CLUSTERED
+  , DeptName VARCHAR(50) NOT NULL
+  , SysStartTime DATETIME2 GENERATED ALWAYS AS ROW START NOT NULL
+  , SysEndTime DATETIME2 GENERATED ALWAYS AS ROW END NOT NULL
+  , PERIOD FOR SYSTEM_TIME (SysStartTime, SysEndTime)
+)
+WITH (SYSTEM_VERSIONING = ON (HISTORY_TABLE = dbo.DepartmentHistory));
+----
+
+This example comes from official documentation about temporal tables: https://docs.microsoft.com/en-us/sql/relational-databases/tables/creating-a-system-versioned-temporal-table?view=sql-server-ver15
+
+Then I add some records to our regular table:
```

Comment:
Why add _"regular"_? If I understand this correctly, the table has been set-up to track history of records, so, wouldn't it make more sense to talk about a _"temporal table"_ instead?

Reply:
I added "regular" to clarify that I wasn't talking about history table, but probably that was indeed redundant